clientapi/routing/consent_tracking.go

```diff
@@ -0,0 +1,226 @@
+				return &internalError
+			}
+			if err = userAPI.QueryPolicyVersion(req.Context(), &userapi.QueryPolicyVersionRequest{
+				LocalPart: localPart,
+			}, res); err != nil {
+				logrus.WithError(err).Error("unable to print consent template")
+				return &internalError
+			}
+			data.HasConsented = res.PolicyVersion == consentCfg.Version
+		}
+
+		err := consentCfg.Templates.ExecuteTemplate(writer, consentCfg.Version+".gohtml", data)
+		if err != nil {
+			logrus.WithError(err).Error("unable to print consent template")
+			return nil
+		}
+		return nil
+	case http.MethodPost:
+		localPart, _, err := gomatrixserverlib.SplitID('@', data.User)
+		if err != nil {
+			logrus.WithError(err).Error("unable to split username")
+			return &internalError
+		}
+
+		ok, err := validHMAC(data.User, data.UserHMAC, consentCfg.FormSecret)
+		if err != nil || !ok {
+			_, err = writer.Write([]byte("invalid HMAC provided"))
+			if err != nil {
+				return &internalError
+			}
+			return &internalError
+		}
+		if err := userAPI.PerformUpdatePolicyVersion(
+			req.Context(),
+			&userapi.UpdatePolicyVersionRequest{
+				PolicyVersion: data.Version,
+				LocalPart:     localPart,
+			},
+			&userapi.UpdatePolicyVersionResponse{},
+		); err != nil {
+			_, err = writer.Write([]byte("unable to update database"))
+			if err != nil {
+				logrus.WithError(err).Error("unable to write to database")
+			}
+			return &internalError
+		}
+		// display the privacy policy without a form
+		data.PublicVersion = false
+		data.HasConsented = true
+
+		err = consentCfg.Templates.ExecuteTemplate(writer, consentCfg.Version+".gohtml", data)
+		if err != nil {
+			logrus.WithError(err).Error("unable to print consent template")
+			return &internalError
+		}
+		return nil
+	}
+	return &util.JSONResponse{Code: http.StatusOK}
+}
+
+func sendServerNoticeForConsent(userAPI userapi.UserInternalAPI, rsAPI api.RoomserverInternalAPI,
+	cfgNotices *config.ServerNotices,
+	cfgClient *config.ClientAPI,
+	senderDevice *userapi.Device,
+	accountsDB userdb.Database,
+	asAPI appserviceAPI.AppServiceQueryAPI,
+) {
+	res := &userapi.QueryOutdatedPolicyResponse{}
+	if err := userAPI.QueryOutdatedPolicy(context.Background(), &userapi.QueryOutdatedPolicyRequest{
+		PolicyVersion: cfgClient.Matrix.UserConsentOptions.Version,
+	}, res); err != nil {
+		logrus.WithError(err).Error("unable to fetch users with outdated consent policy")
+		return
+	}
+
+	var (
+		consentOpts  = cfgClient.Matrix.UserConsentOptions
+		data         = make(map[string]string)
+		err          error
+		sentMessages int
+	)
+
+	if len(res.OutdatedUsers) > 0 {
+		logrus.WithField("count", len(res.OutdatedUsers)).Infof("Sending server notice to users who have not yet accepted the policy")
+	}
+
+	for _, userID := range res.OutdatedUsers {
+		if userID == cfgClient.Matrix.ServerNotices.LocalPart {
+			continue
+		}
+		userID = fmt.Sprintf("@%s:%s", userID, cfgClient.Matrix.ServerName)
```

Comment:
Hang on, we are already dealing with a user ID? Unless you mean `res.OutdatedUsers` are in fact localparts? If so, this needs to be made clear in variable names and field names. "Users" is not a helpful term as it's neither "user ID" nor "localpart".